CPP-900: add pa11y readme file (#254)
Financial-Times/dotcom-tool-kit: plugins/pa11y/readme.md

# @dotcom-tool-kit/pa11y

A plugin to run [Pa11y](https://github.com/pa11y/pa11y) accessibility tests. This plugin uses [Pa11y CI](https://github.com/pa11y/pa11y-ci) to run the tests.

> 

Reviewer: sorry just one last thing following the changes I added in https://github.com/Financial-Times/dotcom-tool-kit/pull/279/files: 

**suggestion**: would it be possible to mention here that the `TEST_URL` environment variable gets populated with the value of the heroku review app, so users can use `process.env.TEST_URL` in their .pa11yci config file?

Reviewer 2: Done 👍 I have made a new section of `## To note` (which definitely needs improving!). Let me know what you think and whether you think it should be somewhere else.

## Installation

With Tool Kit [already set up](https://github.com/financial-times/dotcom-tool-kit#installing-and-using-tool-kit), install this plugin as a dev dependency:

```sh
npm install --save-dev @dotcom-tool-kit/pa11y
```

And add it to your repo's `.toolkitrc.yml`:

```yaml
plugins:
    - '@dotcom-tool-kit/pa11y'
```

## Options

| Key | Description | Default value |
|-|-|-|
| `configFile` | Path to the config file | `.pa11yci.js` |

> ## Hooks

Reviewer: **suggestion**: perhaps we could change this title to `Default Hooks` instead, to make it clear that users can override this hook with another task

Reviewer 3: I prefer something like `Preconfigured Hooks` as it's something you override in your own config rather than something you can modify in the plugin's config...? idk

Reviewer: yes let's go for `Preconfigured Hooks` - it's a more accurate term in this context

> | Event | Description |

Reviewer 3: **nitpick**: we should stick with the `Hook` terminology, rather than introducing new syntax (what do we mean by an 'event'?)

```suggestion
| Hook | Description |
```

|-|-|

> | `test:review` | Runs `pa11y-ci` |

Reviewer: **suggestion**: it might be helpful to also a column for Task, to show that this default hook is running `Pa11y` task

Reviewer 3: **thought**: do we also want to include a description of where the hook will typically be called in a `Hook Description` column or should we leave that to the plugins which actually install the hooks?

Reviewer 2: I am not sure what you mean here, please could you provide an example?

Reviewer: We can use secret squirrel's docs as an [example](https://github.com/Financial-Times/dotcom-tool-kit/tree/main/plugins/secret-squirrel#tasks) - there is a column called Task which mentions that it would run the `SecretSquirrel` [Task](https://github.com/Financial-Times/dotcom-tool-kit/blob/main/plugins/secret-squirrel/src/tasks/secret-squirrel.ts#L7)

# Description 
 Pa11y is your automated accessibility testing pal. It runs accessibility tests on your pages via the command line or Node.js, so you can automate your testing process. Pa11y runs automatically on the `test:review` and `test:staging` hooks, using the information about the app from the state.

# Pa11y documentation
 `https://github.com/pa11y/pa11y`

# Pa11y configurations added on the project
    Pa11ySchema = {

>         - host: Domain against pa11y will run,

Reviewer: suggestion: it would be great to show the type for each option, whether it's an array or string etc.

        - wait: The time in milliseconds to wait before running HTML CodeSniffer on the page,
        - exceptions: An array of result codes and types that you'd like to ignore. You can find the codes for each rule in the console output and the types are error, warning, and notice. Note: warning and notice messages are ignored by default,

>         - hide: A CSS selector to hide elements from testing, selectors can be comma separated. Elements matching this selector will be hidden from testing by styling them with visibility: hidden,

Reviewer: issue: do you mean `hideElements` here? (we might need to change hide to hideElements in the pa11ySchema too)

Reviewer: question: is there a reason why you've reverted `hideElements` back to `hide` in `schema/pa11y.ts`?

>         - viewports: The viewport configuration. This can have any of the properties supported by the puppeteer setViewport method `(https://github.com/puppeteer/puppeteer/blob/main/docs/api.md#pagesetviewportviewport),

Reviewer: issue: this link is broken

        - screenCapturePath: This allows you to capture the screen between other actions, useful to verify that the page looks as you expect before the Pa11y test runs. This action takes the form screen capture <file-path>
    }

# Default config used on the project
    '@dotcom-tool-kit/pa11y':
        host: 'https://ft.com'

# Pa11y implemented output
    When pa11y plugin runs, the output on the terminal will be a list of the issues found, followed for their `errorCode` so they can be reviewed

- Output example: 
        Running Pa11y on `The URL that Pa11y was run against`, document title `The title of the page that was tested`
    
        Issue # of total: 
            {
                code: 'WCAG2AA.Principle1.Guideline1_1.1_1_1.H30.2',
                context: '<a href="https://example.com/"><img src="example.jpg" alt=""/></a>',
                message: 'Img element is the only content of the link, but is missing alt text. The alt text should describe the purpose of the link.',
                selector: 'html > body > p:nth-child(1) > a',
                type: 'error',
                typeCode: 1
            }

>             // more issues appear here

Author: That information was on the first commit, 
![image](https://user-images.githubusercontent.com/100845808/177200769-fd1499f9-02e4-4df1-81b5-5c9b0ec68df0.png)
but Ivo suggested it was not necessary since its already implemented, and its not going to be installed every time its going to be used, 
![image](https://user-images.githubusercontent.com/100845808/177200897-383774b0-45b4-41ff-897f-6442627ae2fe.png)
there is a link to the pa11y documentation where it is described how to install it if neccesary

>             // more issues appear here

Reviewer: what I meant was - it would be good to have instructions in how to install it in toolkit ( instead of installing pa11y from the documentation). something like 
```
With Tool Kit [already set up](https://github.com/financial-times/dotcom-tool-kit#installing-and-using-tool-kit), install this plugin as a dev dependency:

npm install --save-dev @dotcom-tool-kit/secret-squirrel
And add it to your repo's .toolkitrc.yml:

plugins:
  - '@dotcom-tool-kit/secret-squirrel'
 ```

> # Description 

Reviewer 2: **Request:** We should give this document a title that reflects the contents. Here is a suggestion inspired by the recent [Secret Squirrel plugin](https://github.com/Financial-Times/dotcom-tool-kit/tree/main/plugins/secret-squirrel#dotcom-tool-kitsecret-squirrel):

```suggestion
# @dotcom-tool-kit/pa11y
```

>  Pa11y is your automated accessibility testing pal. It runs accessibility tests on your pages via the command line or Node.js, so you can automate your testing process. Pa11y runs automatically on the `test:review` and `test:staging` hooks, using the information about the app from the state.

Reviewer 2: **Request:** The description for the Tool Kit Plugin rather than the tool itself. We should write this description in a way that tells our consumers how this plugin will allow them to use Pa11y with their app via Tool Kit. 

```suggestion
A Tool Kit plugin to run [Pa11y](https://github.com/pa11y/pa11y). This plugin runs automatically on the `test:review` and `test:staging` hooks, using the information about the app from the state.
```

Reviewer 2: **issue:** If #260 is merged before this PR, we will need to update the documented hooks as #260 removes the default use of the `test:staging` hook: https://github.com/Financial-Times/dotcom-tool-kit/pull/260/commits/dea7d5a61dcc7cde81e6c09fd44667b339948cbf

# Pa11y documentation

>  `https://github.com/pa11y/pa11y`

Reviewer 2: **Request:** In my previous request, we move this link to the opening paragraph above. If that is implemented, we can remove this heading and URL.

>  # Installation and usage

Reviewer 2: **Issue:** We should ensure we define the correct [heading levels](https://www.markdownguide.org/basic-syntax/#headings) throughout the document. This will make it easier for the reader to scan the whole document to locate the information they need.

```suggestion
 ## Installation and usage
```

    With Tool Kit [already set up](https://github.com/financial-times/dotcom-tool-kit#installing-and-using-tool-kit), install this plugin as a dev dependency:

    npm install --save-dev @dotcom-tool-kit/pa11y
    And add it to your repo's .toolkitrc.yml:

    plugins:
    - '@dotcom-tool-kit/pa11y'

> 

Reviewer 2: **Issue:** We are missing markdown formatting that will make this section easier to read. Please see the markdown code for the comment below to see the missing formatting.

Reviewer 2: With Tool Kit [already set up](https://github.com/financial-times/dotcom-tool-kit#installing-and-using-tool-kit), install this plugin as a dev dependency:

```
    npm install --save-dev @dotcom-tool-kit/pa11y
```

And add it to your repo's `.toolkitrc.yml`:

```
plugins:
    - '@dotcom-tool-kit/pa11y'
```

# Pa11y configurations added on the project
    Pa11ySchema = {
        - host (string): Domain against pa11y will run,
        - wait (number): The time in milliseconds to wait before running HTML CodeSniffer on the page,
        - exceptions (array.string): An array of result codes and types that you'd like to ignore. You can find the codes for each rule in the console output and the types are error, warning, and notice. Note: warning and notice messages are ignored by default,
        - hideElements (array.string): A CSS selector to hide elements from testing, selectors can be comma separated. Elements matching this selector will be hidden from testing by styling them with visibility: hidden,
        - viewports (array.string): The viewport configuration. This can have any of the properties supported by the puppeteer setViewport method,
        - screenCapturePath (string): This allows you to capture the screen between other actions, useful to verify that the page looks as you expect before the Pa11y test runs. This action takes the form screen capture <file-path>

>     }

Reviewer 2: **Suggestion:** Could we present this as a table? Like the following:

| Value      | Type     | Description                                                                                                                                                                                                                     |
| ---------- | -------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| host       | string   | Domain pa11y will run against                                                                                                                                                                                                   |
| wait       | number   | The time in milliseconds to wait before running HTML CodeSniffer on the page                                                                                                                                                    |
| exceptions | string[] | An array of result codes and types that you'd like to ignore. You can find the codes for each rule in the console output and the types are error, warning, and notice. Note: warning and notice messages are ignored by default |
| etc...     | etc...   | etc...                                                                                                                                                                                                                          |

# Default config used on the project
    '@dotcom-tool-kit/pa11y':

>         host: 'https://ft.com'

Reviewer 2: **Praise:** I agree this is a great addition!

**Request:** Please could this be wrapped in a code block?

# Pa11y implemented output
    When pa11y plugin runs, the output on the terminal will be a list of the issues found, followed for their `errorCode` so they can be reviewed

- Output example: 

>         Running Pa11y on `The URL that Pa11y was run against`, document title `The title of the page that was tested`

Reviewer 2: **Request:** Please could we replace the examples here with something a little more realistic such as:

```suggestion
        Running Pa11y on `https://www.ft.com/example`, document title `Example page | Financial Times`
```

    
        Issue # of total: 
            {
                code: 'WCAG2AA.Principle1.Guideline1_1.1_1_1.H30.2',
                context: '<a href="https://example.com/"><img src="example.jpg" alt=""/></a>',
                message: 'Img element is the only content of the link, but is missing alt text. The alt text should describe the purpose of the link.',
                selector: 'html > body > p:nth-child(1) > a',
                type: 'error',
                typeCode: 1
            }

>             // more issues appear here

Reviewer 2: **Request:** Please could we wrap this in a code block?

# @dotcom-tool-kit/pa11y
 A Tool Kit plugin to run [Pa11y](https://github.com/pa11y/pa11y). This plugin runs accessibility tests on your pages via the command line or Node.js, so you can automate your testing process. Pa11y runs automatically on the `test:review`hook, using the information about the app from the state.

 ## Installation and usage
   With Tool Kit [already set up](https://github.com/financial-times/dotcom-tool-kit#installing-and-using-tool-kit), install this plugin as a dev dependency:

```
    npm install --save-dev @dotcom-tool-kit/pa11y
```

And add it to your repo's `.toolkitrc.yml`:

```
plugins:
    - '@dotcom-tool-kit/pa11y'
```

## Pa11y configurations added on the project
Pa11ySchema:

| Value | Type | Description |
|:---:|:---:|---|
| host | string | Domain against pa11y will run |
| wait | number | The time in milliseconds to wait before running HTML CodeSniffer on the page |
| exceptions | array.string | An array of result codes and types that you'd like to ignore. You can find the codes for each rule in the console output and the types are error, warning, and notice. Note: warning and notice messages are ignored by default |

> | hideElements | array.string | A CSS selector to hide elements from testing, selectors can be comma separated. Elements matching this selector will be hidden from testing by styling them with visibility: hidden |

Reviewer: I was about to approve it, but then saw this - 

**suggestion**: hideElements is `string` instead of `array.string`

    Pa11ySchema:

| Value | Type | Description |
|:---:|:---:|---|
| host | string | Domain against pa11y will run |
| wait | number | The time in milliseconds to wait before running HTML CodeSniffer on the page |
| exceptions | array.string | An array of result codes and types that you'd like to ignore. You can find the codes for each rule in the console output and the types are error, warning, and notice. Note: warning and notice messages are ignored by default |
| hideElements | array.string | A CSS selector to hide elements from testing, selectors can be comma separated. Elements matching this selector will be hidden from testing by styling them with visibility: hidden |
| viewports | array.string | The viewport configuration. This can have any of the properties supported by the puppeteer setViewport method |
| screenCapturePath | string | This allows you to capture the screen between other actions, useful to verify that the page looks as you expect before the Pa11y test runs. This action takes the form screen capture |

    

# Default config used on the project
``` 
    '@dotcom-tool-kit/pa11y':
       host: 'https://ft.com'
``` 

# Pa11y implemented output

>     When pa11y plugin runs, the output on the terminal will be a list of the issues found, followed for their `errorCode` so they can be reviewed

Reviewer: todo: can we please unwrap this in a code block?

# Description 

>  Pa11y is your automated accessibility testing pal. It runs accessibility tests on your pages via the command line or Node.js, so you can automate your testing process.

Reviewer 3: **suggestion**: we should go into more detail how Pa11y integrates with Tool Kit here

>  Pa11y requires Node.js 12+ to run

Reviewer 3: **nitpick**: Tool Kit requires Node 14+ so this isn't relevant

# Pa11y documentation

>  `https://github.com/pa11y/pa11y#javascript-interface`

Reviewer 3: **nitpick**: consumers of the Tool Kit plugin won't be particularly concerned about the Javascript interface, rather a general view of what Pa11y does and how to configure it

```suggestion
 `https://github.com/pa11y/pa11y`
```

> 
> # instalation
>   `npm install pa11y`
> 
> # Options
> 
>     -V, --version                  output the version number
>     -n, --environment              output details about the environment Pa11y will run in
>     -s, --standard <name>          the accessibility standard to use: WCAG2A, WCAG2AA (default), WCAG2AAA – only used by htmlcs runner
>     -r, --reporter <reporter>      the reporter to use: cli (default), csv, json
>     -e, --runner <runner>          the test runners to use: htmlcs (default), axe
>     -l, --level <level>            the level of issue to fail on (exit with code 2): error, warning, notice
>     -T, --threshold <number>       permit this number of errors, warnings, or notices, otherwise fail with exit code 2
>     -i, --ignore <ignore>          types and codes of issues to ignore, a repeatable value or separated by semi-colons
>     --include-notices              Include notices in the report
>     --include-warnings             Include warnings in the report
>     -R, --root-element <selector>  a CSS selector used to limit which part of a page is tested
>     -E, --hide-elements <hide>     a CSS selector to hide elements from testing, selectors can be comma separated
>     -c, --config <path>            a JSON or JavaScript config file
>     -t, --timeout <ms>             the timeout in milliseconds
>     -w, --wait <ms>                the time to wait before running tests in milliseconds
>     -d, --debug                    output debug messages
>     -S, --screen-capture <path>    a path to save a screen capture of the page to
>     -A, --add-rule <rule>          WCAG 2.1 rules to include, a repeatable value or separated by semi-colons – only used by htmlcs runner
>     -h, --help                     output usage information
> 
> # Running Tests
> 
>     Run an accessibility test against a URL:
> 
>     pa11y https://example.com
>     Run an accessibility test against a file (absolute paths only, not relative):
> 
>     pa11y ./path/to/your/file.html
>     Run a test with CSV reporting and save to a file:
> 
>     pa11y --reporter csv https://example.com > report.csv
> 
> # Exit Codes
> 
>     The command-line tool uses the following exit codes:
> 
>     0: Pa11y ran successfully, and there are no errors
>     1: Pa11y failed run due to a technical fault
>     2: Pa11y ran successfully but there are errors in the page
>     By default, only accessibility issues with a type of error will exit with a code of 2. This is configurable with the --level flag which can be set to one of the following:
> 
>     error: exit with a code of 2 on errors only, exit with a code of 0 on warnings and notices
>     warning: exit with a code of 2 on errors and warnings, exit with a code of 0 on notices
>     notice: exit with a code of 2 on errors, warnings, and notices
>     none: always exit with a code of 0
> 
> # Tutorials and articles
>     You can find some useful tutorials and articles in the Tutorials section of pa11y.org.
> 
> # Common Questions and Troubleshooting
>     See our Troubleshooting guide at `https://github.com/pa11y/pa11y/blob/master/TROUBLESHOOTING.md` to get the answers to common questions about Pa11y, along with some ideas to help you troubleshoot any problems.

Reviewer 3: **issue**: this documentation is not relevant when using Tool Kit. we shouldn't be copy+pasting other people's documentation anyway.

```suggestion
```

# Pa11y configurations added on the project
    Pa11ySchema = {

>         host: 'string?',

Reviewer 3: **suggestion**: rather than just giving the type we should explain what each option does

        wait: 'number?',
        tests: 'array.string?',
        exceptions: 'array.string?',
        hide: 'array.string?',
        viewports: 'array.string?',
        screenCapturePath: 'string?'
    }

# Default config used on the project
    '@dotcom-tool-kit/pa11y':

>         host: 'https://ft.com'

Reviewer 3: **praise**: good idea to include this!

# Pa11y implemented output

>     When pa11y plugin runs, the output on the terminal will be a list of the issues found, followed for their `errorCode` so they can be reviewed

Reviewer 3: **suggestion**: it would be good to provide a example of the output
